File: smartdevices-app/smartdevicesapp.model/src/main/java/de/vogler_engineering/smartdevicesapp/model/entities/ui/UiAction.java
/**
 * Copyright (c) Vogler Engineering GmbH. All rights reserved.
 * Licensed under the MIT License. See LICENSE.md in the project root for license information.
 */
package de.vogler_engineering.smartdevicesapp.model.entities.ui;

import java.util.Map;

import lombok.Data;
import lombok.EqualsAndHashCode;

@Data
@EqualsAndHashCode(callSuper = true)
public class UiAction extends UiComponent {
    private String jobKey;

    public static UiAction fromComponent(UiComponent element) {
        if(!element.getAdditionalProperties().containsKey("JobKey") &&
                !element.getAdditionalProperties().containsKey("jobKey"))
            return null;

        UiAction action = new UiAction();
        action.setId(element.getId());
        action.setName(element.getName());
        action.setType(element.getType());

        for (Map.Entry<String, Object> entry :
                element.getAdditionalProperties().entrySet()) {
            if(entry.getKey().equalsIgnoreCase("JobKey")){
                action.setJobKey((String) entry.getValue());
            }else {
                action.setAdditionalProperty(entry.getKey(), entry.getValue());
            }
        }
        return action;
    }
}

File: smartdevices-gateway/SmartDevicesGateway.Model/Ui/UiAction.cs
//
// Copyright (c) Vogler Engineering GmbH. All rights reserved.
// Licensed under the MIT License. See LICENSE.md in the project root for license information.
//
using System.Diagnostics.CodeAnalysis;

namespace SmartDevicesGateway.Model.Ui
{
    public class UiAction : UiComponent
    {
        /// <summary>
        /// Target job of this Action
        /// </summary>
        public string JobKey { get; set; }

        public UiAction()
        {
            Tab = "actions";
        }

        [SuppressMessage("ReSharper", "InvertIf")]
        public UiAction(UiComponent other) : base(other)
        {
            if (other is UiAction otherAction)
            {
                JobKey = otherAction.JobKey;
                return;
            }

            const string key = nameof(JobKey);
            if (AdditionalProperties.ContainsKey(key))
            {
                JobKey = AdditionalProperties[key]?.ToString();
                AdditionalProperties.Remove(key);
            }
        }
    }
}
